lib/firestore/models.dart

```diff
@@ -200,6 +251,138 @@ class Page<T> extends ListBase<T> {
   }
 }
 
+class QueryReference extends Reference {
+  final StructuredQuery _structuredQuery = StructuredQuery();
+
+  QueryReference(FirestoreGateway gateway, String path) : super(gateway, path) {
+    _structuredQuery
+      ..from.add(StructuredQuery_CollectionSelector()..collectionId = id);
+  }
+
+  QueryReference where(
+    String fieldPath, {
+    dynamic isEqualTo,
+    dynamic isLessThan,
+    dynamic isLessThanOrEqualTo,
+    dynamic isGreaterThan,
+    dynamic isGreaterThanOrEqualTo,
+    dynamic arrayContains,
+    List<dynamic> arrayContainsAny,
+    List<dynamic> whereIn,
+    bool isNull = false,
+  }) {
+    if (isEqualTo != null) {
+      _addFilter(fieldPath, isEqualTo,
+          operator: StructuredQuery_FieldFilter_Operator.EQUAL);
+    }
+    if (isLessThan != null) {
+      _addFilter(fieldPath, isLessThan,
+          operator: StructuredQuery_FieldFilter_Operator.LESS_THAN);
+    }
+    if (isLessThanOrEqualTo != null) {
+      _addFilter(fieldPath, isLessThanOrEqualTo,
+          operator: StructuredQuery_FieldFilter_Operator.LESS_THAN_OR_EQUAL);
+    }
+    if (isGreaterThan != null) {
+      _addFilter(fieldPath, isGreaterThan,
+          operator: StructuredQuery_FieldFilter_Operator.GREATER_THAN);
+    }
+    if (isGreaterThanOrEqualTo != null) {
+      _addFilter(fieldPath, isGreaterThanOrEqualTo,
+          operator: StructuredQuery_FieldFilter_Operator.GREATER_THAN_OR_EQUAL);
+    }
+    if (arrayContains != null) {
+      _addFilter(fieldPath, arrayContains,
+          operator: StructuredQuery_FieldFilter_Operator.ARRAY_CONTAINS);
+    }
+    if (arrayContainsAny != null) {
+      _addFilter(fieldPath, arrayContainsAny,
+          operator: StructuredQuery_FieldFilter_Operator.ARRAY_CONTAINS_ANY);
+    }
+    if (whereIn != null) {
+      _addFilter(fieldPath, whereIn,
+          operator: StructuredQuery_FieldFilter_Operator.IN);
+    }
+    if (isNull) {
+      _addFilter(fieldPath, null);
+    }
+
+    return this;
+  }
+
+  /// Returns [QueryReference] that's additionally sorted by the specified
+  /// [fieldPath].
+  ///
+  /// The field is a [String] representing a single field name.
+  /// After a [QueryReference] order by call, you cannot add any more [orderBy]
+  /// calls.
+  QueryReference orderBy(
+    String fieldPath, {
+    bool descending = false,
+  }) {
+    final order = StructuredQuery_Order();
+    order.field_1 = StructuredQuery_FieldReference()..fieldPath = fieldPath;
+    order.direction = descending
+        ? StructuredQuery_Direction.DESCENDING
+        : StructuredQuery_Direction.ASCENDING;
+    _structuredQuery.orderBy.add(order);
+    return this;
+  }
+
+  /// Returns [QueryReference] that's additionally limited to only return up
+  /// to the specified number of documents.
+  QueryReference limit(int count) {
+    _structuredQuery.limit = Int32Value()..value = count;
+    return this;
+  }
```

Comment:
Do these two methods make sense on normal collection requests? At least the `limit` seems weird by itself since you can already specify a limit in `Collection.get()`

Reply:
limit() here is for filters specifically. It is in the official firebase sdk as well and its usage is heavily documented.

If you have an absolutely massive list of documents that the filter lets through, you dont want all of them to be downloaded to the client. This limit stops the query as soon as it hits its target quota.

Reply:
Ok I'm not very familiar with that part of the SDK.